### ISSUE TITLE
TypeErrors: infoHashes is null

### ISSUE THREAD

**Issue by alxhotel** [feross/bittorrent-tracker#205]
A few releases ago, probably since `v8.3.0`, two TypeErrors have appeared (my bad for not reporting this before):

```
var index = peer.socket.infoHashes.indexOf(this.infoHash)
                                      ^

TypeError: Cannot read property 'indexOf' of null
    at Swarm._onAnnounceStopped (/fakepath/tracker/node_modules/bittorrent-tracker/lib/server/swarm.js:106:39)
    at LRU.<anonymous> (/fakepath/tracker/node_modules/bittorrent-tracker/lib/server/swarm.js:31:10)
    at emitOne (events.js:96:13)
    at LRU.emit (events.js:188:7)
    at LRU._checkAge (/fakepath/tracker/node_modules/lru/index.js:104:10)
    at LRU.peek (/fakepath/tracker/node_modules/lru/index.js:64:13)
    at Swarm._getPeers (/fakepath/tracker/node_modules/bittorrent-tracker/lib/server/swarm.js:149:27)
    at Swarm.announce (/fakepath/tracker/node_modules/bittorrent-tracker/lib/server/swarm.js:65:17)
    at announce (/fakepath/tracker/node_modules/bittorrent-tracker/server.js:684:11)
    at /fakepath/tracker/node_modules/bittorrent-tracker/server.js:654:7
```

And this one:

```
if (socket.infoHashes.indexOf(params.info_hash) === -1) {
                           ^

TypeError: Cannot read property 'indexOf' of null
    at /fakepath/tracker/node_modules/bittorrent-tracker/server.js:513:28
    at /fakepath/tracker/node_modules/bittorrent-tracker/server.js:716:7
    at Swarm.announce (/fakepath/tracker/node_modules/bittorrent-tracker/lib/server/swarm.js:62:3)
    at announce (/fakepath/tracker/node_modules/bittorrent-tracker/server.js:684:11)
    at /fakepath/tracker/node_modules/bittorrent-tracker/server.js:654:7
    at /fakepath/tracker/node_modules/bittorrent-tracker/server.js:365:5
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

Seems like it is trying to access the `infoHashes` of a peer already evicted.

I would like to fix them myself, but, unfortunately, I don't have too much time these days.

@feross @DiegoRBaquero @yciabaud

**Comment by DiegoRBaquero** [comment]
@alxhotel Do you have latest version?

**Comment by alxhotel** [comment]
Yep.

`v8.5.0` of `bittorrent-tracker`.
`v6.9.5` of `NodeJS`.

**Comment by DiegoRBaquero** [comment]
@feross https://github.com/feross/bittorrent-tracker/pull/198#discussion_r99274685

This is why closing the socket must be done in nextTick 😛 

When requesting the peer from the swarm, we evict and close its socket but we use it later on.

How I did it in uwt: https://github.com/DiegoRBaquero/uWebTorrentTracker/blob/master/lib/swarm.js#L89

Gonna PR.

**Comment by feross** [comment]
@DiegoRBaquero I'm not sure that adding a `nextTick` is enough to completely solve this issue. Also, it's a bit hacky.

I think we should just ensure that we don't try to use `socket.infoHashes` after the socket is destroyed. PR here: https://github.com/feross/bittorrent-tracker/pull/208

**Comment by feross** [comment]
Released as 8.5.1
